clients/web/src/views/widgets/MarkdownWidget.js

```diff
@@ -4,7 +4,140 @@
 girder.views.MarkdownWidget = girder.View.extend({
     events: {
         'show.bs.tab .g-preview-link': function () {
-            girder.renderMarkdown(this.val(), this.$('.g-markdown-preview'));
+            girder.renderMarkdown(this.val().trim() || 'Nothing to show',
+                                  this.$('.g-markdown-preview'));
+        },
+
+        'shown.bs.tab .g-write-link': function () {
+            this.$('.g-markdown-text').focus();
+        },
+
+        'click .g-browse': function () {
+            this.$('.g-file-input').click();
+        },
+
+        'change .g-file-input': function (e) {
+            var files = e.target.files;
+
+            if (files.length) {
+                this.files = files;
+                this.filesAdded();
+            }
+        },
+
+        'dragenter .g-markdown-drop-zone': function (e) {
+            e.stopPropagation();
+            e.preventDefault();
+            e.originalEvent.dataTransfer.dropEffect = 'copy';
+            this.$('.g-markdown-drop-zone,.g-upload-footer').addClass('dragover');
+        },
+
+        'dragleave .g-markdown-drop-zone': function (e) {
+            e.stopPropagation();
+            e.preventDefault();
+            this.$('.g-markdown-text,.g-upload-footer').removeClass('dragover');
+        },
+
+        'dragover .g-markdown-drop-zone': function (e) {
+            e.originalEvent.dataTransfer.dropEffect = 'copy';
+            this.$('.g-markdown-text,.g-upload-footer').addClass('dragover');
+        },
+
+        'drop .g-markdown-drop-zone': function (e) {
+            e.stopPropagation();
+            e.preventDefault();
+            this.$('.g-markdown-text,.g-upload-footer').removeClass('dragover');
+            this.files = e.originalEvent.dataTransfer.files;
+            this.filesAdded();
+        }
+    },
+
```

Comment:
None of the drag-and-drop functionality has coverage (neither here nor our existing drag-and-drop).  A short time ago I read that we should be able to test this in phantom; perhaps create a new issue to test drag-and-drop code.